File: android/src/main/java/com/reactlibrary/QonversionModule.java
package com.reactlibrary;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import org.json.JSONException;
import java.util.HashMap;
import java.util.Map;
import java.util.List;
import android.app.Application;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.qonversion.sandwich.QonversionEventsListener;
import io.qonversion.sandwich.QonversionSandwich;
import io.qonversion.sandwich.SandwichError;

public class QonversionModule extends ReactContextBaseJavaModule implements QonversionEventsListener {

    private final QonversionSandwich qonversionSandwich;

    private static final String EVENT_ENTITLEMENTS_UPDATED = "entitlements_updated";

    private DeviceEventManagerModule.RCTDeviceEventEmitter eventEmitter = null;

    public QonversionModule(ReactApplicationContext reactContext) {
        super(reactContext);

        qonversionSandwich = new QonversionSandwich(
                (Application) reactContext.getApplicationContext(),
                QonversionModule.this::getCurrentActivity,
                this
        );
    }

    @Override
    public void initialize() {
        super.initialize();

        eventEmitter = getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class);
    }

    @NonNull
    @Override
    public String getName() {
        return "RNQonversion";
    }

    @ReactMethod
    public void storeSDKInfo(String source, String sdkVersion) {
        qonversionSandwich.storeSdkInfo(source, sdkVersion);
    }

    @ReactMethod
    public void initializeSdk(
            String projectKey,
            String launchModeKey,
            @Nullable String environmentKey,
            @Nullable String entitlementsCacheLifetimeKey,
            @Nullable String proxyUrl,
            boolean kidsMode
    ) {
        qonversionSandwich.initialize(
                getReactApplicationContext(),
                projectKey,
                launchModeKey,
                environmentKey,
                entitlementsCacheLifetimeKey,
                proxyUrl,
                kidsMode
        );
    }

    @ReactMethod
    public void syncHistoricalData() {
        qonversionSandwich.syncHistoricalData();
    }

    @ReactMethod
    public void purchase(
            String productId,
            @Nullable String offerId,
            @Nullable Boolean applyOffer,
            @Nullable String oldProductId,
            @Nullable String updatePolicyKey,
            @Nullable ReadableArray contextKeys,
            final Promise promise) {
        List<String> contextKeysList = null;
        if (contextKeys != null) {
            contextKeysList = EntitiesConverter.convertArrayToStringList(contextKeys);
        }
        qonversionSandwich.purchase(productId, offerId, applyOffer, oldProductId, updatePolicyKey, contextKeysList, Utils.getResultListener(promise));
    }

    @ReactMethod
    public void updatePurchase(
            String productId,
            @Nullable String offerId,
            @Nullable Boolean applyOffer,
            String oldProductId,
            @Nullable String updatePolicyKey,
            @Nullable ReadableArray contextKeys,
            final Promise promise
    ) {
        purchase(productId, offerId, applyOffer, oldProductId, updatePolicyKey, contextKeys, promise);
    }

    @ReactMethod
    public void setDefinedProperty(String key, String value) {
        qonversionSandwich.setDefinedProperty(key, value);
    }

    @ReactMethod
    public void setCustomProperty(String key, String value) {
        qonversionSandwich.setCustomProperty(key, value);
    }

    @ReactMethod
    public void userProperties(final Promise promise) {
        qonversionSandwich.userProperties(Utils.getResultListener(promise));
    }

    @ReactMethod
    public void addAttributionData(ReadableMap map, String provider) {
        try {
            HashMap<String, Object> attributesHashMap = EntitiesConverter.convertReadableMapToHashMap(map);
            qonversionSandwich.addAttributionData(provider, attributesHashMap);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    @ReactMethod
    public void checkEntitlements(final Promise promise) {
        qonversionSandwich.checkEntitlements(Utils.getResultListener(promise));
    }

    @ReactMethod
    public void products(final Promise promise) {
        qonversionSandwich.products(Utils.getResultListener(promise));
    }

    @ReactMethod
    public void offerings(final Promise promise) {
        qonversionSandwich.offerings(Utils.getResultListener(promise));
    }

    @ReactMethod
    public void checkTrialIntroEligibilityForProductIds(ReadableArray ids, final Promise promise) {
        final List<String> idList = EntitiesConverter.convertArrayToStringList(ids);
        qonversionSandwich.checkTrialIntroEligibility(idList, Utils.getResultListener(promise));
    }

    @ReactMethod
    public void restore(final Promise promise) {
        qonversionSandwich.restore(Utils.getResultListener(promise));
    }

    @ReactMethod
    public void syncPurchases() {
        qonversionSandwich.syncPurchases();
    }

    @ReactMethod
    public void identify(final String userID, final Promise promise) {
        qonversionSandwich.identify(userID, Utils.getResultListener(promise));
    }

    @ReactMethod
    public void logout() {
        qonversionSandwich.logout();
    }

    @ReactMethod
    public void userInfo(final Promise promise) {
        qonversionSandwich.userInfo(Utils.getResultListener(promise));
    }

    @ReactMethod
    public void remoteConfig(@Nullable final String contextKey, final Promise promise) {
        qonversionSandwich.remoteConfig(contextKey, Utils.getResultListener(promise));
    }

    @ReactMethod
    public void remoteConfigList(final Promise promise) {
        qonversionSandwich.remoteConfigList(Utils.getResultListener(promise));
    }

    @ReactMethod
    public void remoteConfigListForContextKeys(final ReadableArray contextKeys, final boolean includeEmptyContextKey, final Promise promise) {
        final List<String> keysList = EntitiesConverter.convertArrayToStringList(contextKeys);
        qonversionSandwich.remoteConfigList(keysList, includeEmptyContextKey, Utils.getResultListener(promise));
    }

    @ReactMethod
    public void attachUserToExperiment(final String experimentId, final String groupId, final Promise promise) {
        qonversionSandwich.attachUserToExperiment(experimentId, groupId, Utils.getResultListener(promise));
    }

    @ReactMethod
    public void detachUserFromExperiment(final String experimentId, final Promise promise) {
        qonversionSandwich.detachUserFromExperiment(experimentId, Utils.getResultListener(promise));
    }

    @ReactMethod
    public void attachUserToRemoteConfiguration(final String remoteConfigurationId, final Promise promise) {
        qonversionSandwich.attachUserToRemoteConfiguration(remoteConfigurationId, Utils.getResultListener(promise));
    }

    @ReactMethod
    public void detachUserFromRemoteConfiguration(final String remoteConfigurationId, final Promise promise) {
        qonversionSandwich.detachUserFromRemoteConfiguration(remoteConfigurationId, Utils.getResultListener(promise));
    }

    @ReactMethod
    public void isFallbackFileAccessible(final Promise promise) {
        qonversionSandwich.isFallbackFileAccessible(Utils.getResultListener(promise));
    }

    @Override
    public void onEntitlementsUpdated(@NonNull Map<String, ?> map) {
        final WritableMap payload = EntitiesConverter.convertMapToWritableMap(map);
        if (eventEmitter != null) {
            eventEmitter.emit(EVENT_ENTITLEMENTS_UPDATED, payload);
        }
    }
}

File: android/src/main/java/com/qonversion/reactnativesdk/QonversionModule.kt
package com.reactlibrary

import android.app.Activity
import android.app.Application
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.module.annotations.ReactModule
import com.qonversion.reactnativesdk.EntitiesConverter
import com.qonversion.reactnativesdk.NativeQonversionModuleSpec
import com.qonversion.reactnativesdk.Utils.Companion.getResultListener
import io.qonversion.sandwich.ActivityProvider
import io.qonversion.sandwich.BridgeData
import io.qonversion.sandwich.QonversionEventsListener
import io.qonversion.sandwich.QonversionSandwich
import org.json.JSONException

@ReactModule(name = QonversionModule.NAME)
class QonversionModule(reactContext: ReactApplicationContext) : NativeQonversionModuleSpec(reactContext), QonversionEventsListener {
    private val qonversionSandwich = QonversionSandwich(
        (reactContext.applicationContext as Application),
        object : ActivityProvider {
            override val currentActivity: Activity?
                get() = this@QonversionModule.currentActivity
        },
        this
    )

    override fun getName(): String {
        return NAME
    }

    @ReactMethod
    override fun storeSDKInfo(source: String, sdkVersion: String) {
        qonversionSandwich.storeSdkInfo(source, sdkVersion)
    }

    @ReactMethod
    override fun initializeSdk(
        projectKey: String,
        launchModeKey: String,
        environmentKey: String?,
        entitlementsCacheLifetimeKey: String?,
        proxyUrl: String?,
        kidsMode: Boolean
    ) {
        qonversionSandwich.initialize(
            reactApplicationContext,
            projectKey,
            launchModeKey,
            environmentKey,
            entitlementsCacheLifetimeKey,
            proxyUrl,
            kidsMode
        )
    }

    @ReactMethod
    override fun syncHistoricalData() {
        qonversionSandwich.syncHistoricalData()
    }

    @ReactMethod
    override fun purchase(
        productId: String,
        quantity: Double,
        contextKeys: ReadableArray?,
        promoOffer: ReadableMap?,
        offerId: String?,
        applyOffer: Boolean,
        oldProductId: String?,
        updatePolicyKey: String?,
        promise: Promise
    ) {
        var contextKeysList: List<String>? = null
        if (contextKeys != null) {
            contextKeysList = EntitiesConverter.convertArrayToStringList(contextKeys)
        }
        qonversionSandwich.purchase(
            productId,
            offerId,
            applyOffer,
            oldProductId,
            updatePolicyKey,
            contextKeysList,
            getResultListener(promise)
        )
    }

    @ReactMethod
    override fun updatePurchase(
        productId: String,
        offerId: String?,
        applyOffer: Boolean,
        oldProductId: String?,
        updatePolicyKey: String?,
        promise: Promise
    ) {
        purchase(
            productId,
            1.0,
            null,
            null,
            offerId,
            applyOffer,
            oldProductId,
            updatePolicyKey,
            promise
        )
    }

    @ReactMethod
    override fun setDefinedProperty(key: String, value: String) {
        qonversionSandwich.setDefinedProperty(key, value)
    }

    @ReactMethod
    override fun setCustomProperty(key: String, value: String) {
        qonversionSandwich.setCustomProperty(key, value)
    }

    @ReactMethod
    override fun userProperties(promise: Promise) {
        qonversionSandwich.userProperties(getResultListener(promise))
    }

    @ReactMethod
    override fun addAttributionData(map: ReadableMap?, provider: String) {
        try {
            val attributesHashMap = EntitiesConverter.convertReadableMapToHashMap(
                map!!
            )
            qonversionSandwich.addAttributionData(provider, attributesHashMap)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    @ReactMethod
    override fun checkEntitlements(promise: Promise) {
        qonversionSandwich.checkEntitlements(getResultListener(promise))
    }

    @ReactMethod
    override fun products(promise: Promise) {
        qonversionSandwich.products(getResultListener(promise))
    }

    @ReactMethod
    override fun offerings(promise: Promise) {
        qonversionSandwich.offerings(getResultListener(promise))
    }

    @ReactMethod
    override fun checkTrialIntroEligibilityForProductIds(ids: ReadableArray?, promise: Promise) {
        val idList = EntitiesConverter.convertArrayToStringList(
            ids!!
        )
        qonversionSandwich.checkTrialIntroEligibility(idList, getResultListener(promise))
    }

    @ReactMethod
    override fun restore(promise: Promise) {
        qonversionSandwich.restore(getResultListener(promise))
    }

    @ReactMethod
    override fun syncPurchases() {
        qonversionSandwich.syncPurchases()
    }

    @ReactMethod
    override fun identify(userId: String, promise: Promise) {
        qonversionSandwich.identify(userId, getResultListener(promise))
    }

    @ReactMethod
    override fun logout() {
        qonversionSandwich.logout()
    }

    @ReactMethod
    override fun userInfo(promise: Promise) {
        qonversionSandwich.userInfo(getResultListener(promise))
    }

    @ReactMethod
    override fun remoteConfig(contextKey: String?, promise: Promise) {
        qonversionSandwich.remoteConfig(contextKey, getResultListener(promise))
    }

    @ReactMethod
    override fun remoteConfigList(promise: Promise) {
        qonversionSandwich.remoteConfigList(getResultListener(promise))
    }

    @ReactMethod
    override fun remoteConfigListForContextKeys(
        contextKeys: ReadableArray?,
        includeEmptyContextKey: Boolean,
        promise: Promise
    ) {
        val keysList = EntitiesConverter.convertArrayToStringList(
            contextKeys!!
        )
        qonversionSandwich.remoteConfigList(
            keysList,
            includeEmptyContextKey,
            getResultListener(promise)
        )
    }

    @ReactMethod
    override fun attachUserToExperiment(experimentId: String, groupId: String, promise: Promise) {
        qonversionSandwich.attachUserToExperiment(experimentId, groupId, getResultListener(promise))
    }

    @ReactMethod
    override fun detachUserFromExperiment(experimentId: String, promise: Promise) {
        qonversionSandwich.detachUserFromExperiment(experimentId, getResultListener(promise))
    }

    @ReactMethod
    override fun attachUserToRemoteConfiguration(remoteConfigurationId: String, promise: Promise) {
        qonversionSandwich.attachUserToRemoteConfiguration(
            remoteConfigurationId,
            getResultListener(promise)
        )
    }

    @ReactMethod
    override fun detachUserFromRemoteConfiguration(remoteConfigurationId: String, promise: Promise) {
        qonversionSandwich.detachUserFromRemoteConfiguration(
            remoteConfigurationId,
            getResultListener(promise)
        )
    }

    @ReactMethod
    override fun isFallbackFileAccessible(promise: Promise) {
        qonversionSandwich.isFallbackFileAccessible(getResultListener(promise))
    }

    override fun syncStoreKit2Purchases() {
        // iOS only
    }

    override fun collectAdvertisingId() {
        // iOS only
    }

    override fun collectAppleSearchAdsAttribution() {
        // iOS only
    }

    override fun presentCodeRedemptionSheet() {
        // iOS only
    }

    override fun promoPurchase(productId: String?, promise: Promise?) {
        // iOS only
    }

    override fun getPromotionalOffer(product: String?, discount: String?, promise: Promise?) {
        // iOS only
    }

    override fun onEntitlementsUpdated(entitlements: BridgeData) {
        val mappedEntitlements = EntitiesConverter.convertMapToWritableMap(entitlements)
        emitOnEntitlementsUpdated(mappedEntitlements)
    }

    companion object {
        const val NAME = "RNQonversion"
    }
}
